### PR TITLE
Support Literal types in script runner

### LINKED ISSUE #1173: Support Literal types in script runner
<h2>Pre-bug-report checklist</h2>

**1. This bug can be reproduced using pure Argo YAML**
- [ ] Yes [👉 Please report a bug to the Argo Workflows GitHub 👈](https://github.com/argoproj/argo-workflows/issues/new/choose)
- [X] No

_If yes, it is more likely to be an Argo bug unrelated to Hera. Please double check before submitting an issue to Hera._

**2. This bug occurs in Hera when...**
- [ ] exporting to YAML
- [ ] submitting to Argo
- [X] running on Argo with the Hera runner
- [ ] other: 

<h2>Bug report</h2>

**Describe the bug**
_A clear and concise description of what the bug is:_

Using a function with a Literal argument in the runner results in the following trace:

_Error log if applicable_:
```
Traceback (most recent call last):
  File "/.../python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File ".../python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File ".../python3.9/site-packages/hera/workflows/runner.py", line 6, in <module>
    _run()
  File "/.../python3.9/site-packages/hera/workflows/_runner/util.py", line 254, in _run
    result = _runner(args.entrypoint, kwargs_list)
  File ".../python3.9/site-packages/hera/workflows/_runner/util.py", line 228, in _runner
    return function(**kwargs)
  File ".../python3.9/site-packages/hera/workflows/_runner/util.py", line 58, in inner
    filtered_kwargs = {key: _parse(value, key, f) for key, value in kwargs.items() if _is_kwarg_of(key, f)}
  File ".../python3.9/site-packages/hera/workflows/_runner/util.py", line 58, in <dictcomp>
    filtered_kwargs = {key: _parse(value, key, f) for key, value in kwargs.items() if _is_kwarg_of(key, f)}
  File ".../python3.9/site-packages/hera/workflows/_runner/util.py", line 89, in _parse
    if _is_str_kwarg_of(key, f) or _is_artifact_loaded(key, f) or _is_output_kwarg(key, f):
  File ".../python3.9/site-packages/hera/workflows/_runner/util.py", line 142, in _is_str_kwarg_of
    return issubclass(type_, str)
TypeError: issubclass() arg 1 must be a class
```

**To Reproduce**
Full Hera code to reproduce the bug:
```py
from typing import Literal
from hera.workflows import Workflow, script

@script(constructor="runner", image="my-image")
def literal_str(my_str: Literal["foo", "bar"]):
    print(my_str)

with Workflow(name="my-workflow", entrypoint="steps") as w:
    with Steps(name="steps"):
        literal_str(arguments={"my_str": "foo"})
```

**Expected behavior**
_A clear and concise description of what you expected to happen:_
* For unsupported types, the error should indicate which argument had the issue, and that it is due to an unsupported type; at the moment, the key is lost, and the error message is not obvious.
* Ideally:
  - no error
  - `_is_str_kwarg_of` should return `True`
  - `Parameter.enum` should be inferred to be `["foo", "bar"]`

**Environment**
- Hera Version: 5.16.2
- Python Version: 3.9.18
- Argo Version: 3.4.16.7

### PR DESCRIPTION
**Pull request by alicederyn**
**Pull Request Checklist**
- [X] Fixes #1173
- [X] Tests added
- [ ] Documentation/examples added
- [X] [Good commit messages](https://cbea.ms/git-commit/) and/or PR title

**Description of PR**
Currently, the script runner does not support `Literal` types, as `origin_type_issubtype` does not special-case them, so `_is_str_kwarg_of` does not return True.

This PR adds supports for Literals to `origin_type_issubtype`, and additionally copies their values from the annotation into the `enum` field on input Parameters. It also combines two paths in `_get_inputs_from_callable` into one by using `construct_io_from_annotation` instead of having a fallback branch if `get_workflow_annotation` returns `None`; this fixes a bug where the default was being verified as None for optional strings only if there was no IO annotation.

### CODE REVIEW

File: src/hera/workflows/script.py [hunk 1]
```diff
@@ -379,6 +380,7 @@ def _get_parameters_from_callable(source: Callable) -> List[Parameter]:
             default = MISSING
 
         param = Parameter(name=p.name, default=default)
+        add_metadata_from_type(param, p.annotation)
```
Comment (alicederyn) [inline src/hera/workflows/script.py:383]: I wasn't sure if we wanted to populate param.enum from the Literal metadata in the non-experimental path, or just when the `script_annotations` experimental feature is enabled, so I put this change into a separate commit to make it easy to revert.

Comment (sambhav) [reply src/hera/workflows/script.py:383]: I would prefer to do it just for script annotations for now

Comment (alicederyn) [reply src/hera/workflows/script.py:383]: That does mean that the generated YAML for a non-Annotated script will change depending on whether an unrelated experimental feature is enabled though.

File: src/hera/shared/_type_util.py [hunk 2]
```diff
@@ -82,6 +83,13 @@ def get_workflow_annotation(annotation: Any) -> Optional[Union[Artifact, Paramet
     return metadata[0]
 
 
+def add_metadata_from_type(parameter: Parameter, annotation: Any) -> None:
```
Comment (jeongukjae) [inline src/hera/shared/_type_util.py:86]: should be great if docstring is added here

File: src/hera/shared/_type_util.py [hunk 3]
```diff
@@ -91,13 +99,19 @@ def construct_io_from_annotation(python_name: str, annotation: Any) -> Union[Par
...
-        return annotation_copy
+        io = workflow_annotation.copy()
+    else:
+        io = Parameter()
+
+    io.name = io.name or python_name
+    if isinstance(io, Parameter) and not io.enum:
+        type_ = unwrap_annotation(annotation)
+        if get_origin(type_) is Literal:
+            io.enum = list(get_args(type_))
```
Comment (jeongukjae) [inline src/hera/shared/_type_util.py:112]: seems we can reuse above function

```suggestion
    if isinstance(io, Parameter):
        add_metadata_from_type(io, annotation)
```

File: src/hera/shared/_type_util.py (hunk 2 above)
Comment (jeongukjae) [reply src/hera/shared/_type_util.py:86]: are there any other things you want to do in this function? this function name looks generic.

Comment (alicederyn) [reply src/hera/shared/_type_util.py]: That was the intention, but looking at the Parameter type, I can't see any other fields that we would want to set based on the type annotation, so renamed to be more specific -- thanks!

File: src/hera/shared/_type_util.py (hunk 3 above)
Comment (alicederyn) [reply src/hera/shared/_type_util.py]: Whoops! This was indeed the intention, must have lost it in my git history.... Fixed, thanks!

**Review by jeongukjae — APPROVED** [review]
:+1:

**Review by sambhav — APPROVED** [review]
